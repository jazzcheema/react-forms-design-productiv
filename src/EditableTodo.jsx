import React, { useState } from "react";
import Todo from "./Todo";
import TodoForm from "./TodoForm";

/** Show editable todo item.
 *
 * Props
 * - todo like {id, title, description, priority}
 * - update(): fn to call to update a todo
 * - removeMe(): fn to call to remove this todo
 *
 * EditableTodoList -> EditableTodo -> { Todo, TodoForm }
 */

function EditableTodo({ update, removeMe, todo }) {

  const [isEditing, setIsEditing] = useState(false);

  /** Toggle if this is being edited */
  function toggleEdit() {
    setIsEditing(currState => !currState);
  }

  /** Call removeMe fn passed to this. */
  function handleDelete() {
    removeMe(todo.id);
  }

  /** Edit form saved; toggle isEditing and update in ancestor. */
  function handleSave(formData) {
    update(formData);
    toggleEdit();
  }

  return (
    <div className="EditableTodo">

      {isEditing
        ? (
          <TodoForm
            initialFormData={todo}
            handleSave={handleSave} />
        ) : (
          <div className="mb-3">
            <div className="float-end text-sm-end">
              <button
                className="EditableTodo-toggle btn-link btn btn-sm"
                onClick={toggleEdit}>
                Edit
              </button>
              <button
                className="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                onClick={handleDelete}>
                Del
              </button>
            </div>
            <Todo todo={todo} />
          </div>
        )
      }

    </div>
  );
}

export default EditableTodo;
