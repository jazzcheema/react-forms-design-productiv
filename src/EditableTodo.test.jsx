import { describe, it, expect, vi } from "vitest";

import React from "react";
import { render, fireEvent } from "@testing-library/react";
import EditableTodo from "./EditableTodo";

describe("EditableTodo component", function () {

    const testTodo = {
        title: "Test Title1",
        description: "Test Description1",
        priority: "3"
    };


    function update() {
        return true;
    }

    it("renders without crashing", function () {
        render(<EditableTodo todo={testTodo} update={update} removeMe={() => true} />);
    });

    it("Matches snapshot", function () {
        const { container } = render(<EditableTodo todo={testTodo} update={update} removeMe={() => true} />);
        expect(container).toMatchSnapshot();
    });

    // it("renders form when edit button clicked, renders todo when submit button clicked", function () {


    //     const { container, debug } = render(
    //         <TodoForm
    //             initialFormData={testInitial}
    //             handleSave={handleSave} />);

    //     const titleInput = container.querySelector("#newTodo-title");
    //     const descriptionInput = container.querySelector("#newTodo-description");
    //     const priorityInput = container.querySelector("#newTodo-priority");

    //     fireEvent.change(titleInput, {target: {value: "test title"}})
    //     fireEvent.change(descriptionInput, {target: {value: "test description"}})
    //     fireEvent.change(priorityInput, {target: {value: "1"}})

    //     const submitBtn = container.querySelector(".NewTodoForm-addBtn");
    //     fireEvent.click(submitBtn);

    //     expect(titleInput.textContent).toEqual("");
    //     expect(descriptionInput.textContent).toEqual("");
    //     expect(priorityInput.value).toEqual("3");

    //     expect(formData).toEqual({
    //         title: "test title",
    //         description: "test description",
    //         priority: "1"
    //     })
    // });

    it("calls removeMe when delete button clicked", function() {
        function removeMe() {
            return true;
        }

        let mockRemoveMe = vi.fn(removeMe);

        const { container } = render(<EditableTodo todo={testTodo} update={update} removeMe={mockRemoveMe} />);

        const deleteBtn = container.querySelector(".EditableTodo-delBtn");
        fireEvent.click(deleteBtn);

        expect(mockRemoveMe.mock.calls.length).toEqual(1);
    })

});