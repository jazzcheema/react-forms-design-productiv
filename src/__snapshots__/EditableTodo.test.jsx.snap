// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditableTodo component > Matches snapshot 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <div
      class="mb-3"
    >
      <div
        class="float-end text-sm-end"
      >
        <button
          class="EditableTodo-toggle btn-link btn btn-sm"
        >
          Edit
        </button>
        <button
          class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
        >
          Del
        </button>
      </div>
      <div
        class="Todo"
      >
        <div>
          <b>
            Test Title1
          </b>
           
          <small>
            (priority: 
            3
            )
          </small>
        </div>
        <div>
          <small>
            Test Description1
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoForm component > Matches snapshot 1`] = `
<div>
  <form
    class="NewTodoForm"
  >
    <div
      class="mb-3"
    >
      <input
        aria-label="Title"
        class="form-control"
        id="newTodo-title"
        name="title"
        placeholder="Title"
        required=""
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <textarea
        aria-label="Description"
        class="form-control"
        id="newTodo-description"
        name="description"
        placeholder="Description"
      />
    </div>
    <div
      class="mb-3 d-flex justify-content-between"
    >
      <div
        class="w-75 d-flex justify-content-between"
      >
        <label
          class="d-inline-flex"
          for="newTodo-priority"
        >
          Priority:; 
        </label>
        <select
          class="form-control form-control-sm d-inline-flex"
          id="newTodo-priority"
          name="priority"
        >
          <option
            value="1"
          >
            Ultra-Über
          </option>
          <option
            value="2"
          >
            Über
          </option>
          <option
            value="3"
          >
            Meh
          </option>
        </select>
      </div>
      <button
        class="btn-primary rig btn btn-sm NewTodoForm-addBtn"
      >
        Gø!
      </button>
    </div>
  </form>
</div>
`;
